src/main/java/org/broadinstitute/hellbender/engine/spark/SparkSharder.java

```diff
@@ -146,48 +202,68 @@ public T apply(@Nullable Tuple2<I, Iterable<L>> input) {
-        for (int i = 0; i < locatables.getNumPartitions(); i++) {
-            maxEndPartitionIndexes.add(i);
-        }
         OverlapDetector<PartitionLocatable<SimpleInterval>> overlapDetector = OverlapDetector.create(partitionReadExtents);
-        List<PartitionLocatable<I>> indexedIntervals = new ArrayList<>();
-        int lastPartitionStartIndex = 0;
-        for (I interval : intervals) {
-            int[] partitionIndexes = overlapDetector.getOverlaps(interval).stream()
+        Broadcast<OverlapDetector<PartitionLocatable<SimpleInterval>>> overlapDetectorBroadcast = ctx.broadcast(overlapDetector);
+        JavaRDD<PartitionLocatable<I>> indexedIntervals = intervals.map(interval -> {
+            int[] partitionIndexes = overlapDetectorBroadcast.getValue().getOverlaps(interval).stream()
                     .mapToInt(PartitionLocatable::getPartitionIndex).toArray();
             if (partitionIndexes.length == 0) {
-                // interval does not overlap any partition - add it to the last one
-                indexedIntervals.add(new PartitionLocatable<>(lastPartitionStartIndex, interval));
-                continue;
+                // interval does not overlap any partition - add it to the one after the interval start
+                int i = Collections.binarySearch(firstLocatables, new SimpleInterval(interval), (o1, o2) -> IntervalUtils.compareLocatables(o1, o2, sequenceDictionary));
+                if (i >= 0) {
+                    throw new IllegalStateException(); // TODO: no overlaps, yet start of interval matches a partition read extent start
+                }
+                int insertionPoint = -i - 1;
+                if (insertionPoint == firstLocatables.size()) {
+                    insertionPoint = firstLocatables.size() - 1;
+                }
+                return new PartitionLocatable<>(insertionPoint, interval);
             }
             Arrays.sort(partitionIndexes);
             int startIndex = partitionIndexes[0];
             int endIndex = partitionIndexes[partitionIndexes.length - 1];
-            indexedIntervals.add(new PartitionLocatable<I>(startIndex, interval));
+            return new PartitionLocatable<>(startIndex, endIndex, interval);
+        });
+
+        // Create an RDD of intervals with the same number of partitions as the locatables, and where each interval
+        // is in its start partition. Within each partition, intervals are sorted by IntervalUtils#compareLocatables.
+        JavaRDD<PartitionLocatable<I>> indexedIntervalsRepartitioned = indexedIntervals
+                .mapToPair(interval ->
+                        new Tuple2<>(interval, (Void) null))
+                .repartitionAndSortWithinPartitions(new PartitionLocatablePartitioner(locatables.getNumPartitions()), new PartitionLocatableComparator<I>(sequenceDictionary))
+                .keys();
+
+        indexedIntervalsRepartitioned.cache(); // cache since we need to do two calculations on the intervals
```

Comment:
Are you sure that ends up being faster? When we were doing mark duplicates work we found that caching ended up costing significantly more than just doing the entire tools work twice (regardless of which cache we used). This is probably a reasonable use of it however...

Reply:
It was faster when I was doing some test runs on exome data.